### DISCUSSION TITLE
Support for detecting if an address is unused

### DISCUSSION THREAD

**Discussion by torkelrogstad** [romanz/electrs#920] [Q&A]
I need to determine if an address which I don't control is unused or not. These addresses belong to customers of mine, so their potential usage is completely outside of my control. If I query for the history of the script hash of the address, this quickly balloons into very slow calls. If a call is slow to return I can use that as a heuristic that the address is _probably_ used, but it's not definite. I want to avoid using external providers, as the service I run prides itself on having as few 3rd parties as possible. I'm aware that the Blockstream fork of this project might be better suited, but I really like that this version requires so little resources. 

Would it be an idea to implement a new protocol method, something along the lines of `blockchain.scripthash.is_unused`? Perhaps it's preposterous to suggest a brand new protocol method - and there's a much easier way of accomplishing this. If so, please LMK!

**Comment by antonilol** [comment]
a new protocol method should be added to the spec first, or implemented in some implementation specific extension (like `electrs.blockchain.scripthash.is_unused` or so?) i do not know of an easier way that short-circuits when at least one tx is found other that a new method
also, what do you exactly mean with unused, that there exists no transactions that 'send to that address' (confirmed or unconfirmed?) or that there is no balance (no utxos)? privacy minded wallets wont use an address twice, so addresses that had funds that dont have them anymore will not be used anymore
if you mean "no transaction that sends to that address" this should be easy to implement, and would return "used" as soon as it finds a tx, or "unused" after not finding any (or true/false instead of used/unused)

**Comment by torkelrogstad** [comment]
By "unused" I mean that the address has never received funds. 

I know that privacy minded wallets won't reuse addresses. The issue here is that I interface with the wallets of my customers, and they're not necessarily privacy minded. I therefore have no control over if my customers reuse addresses or not, but I have a need check if a given customer address has transacted before.

**Reply by romanz** [reply]
It is possible to add a new RPC that will lookup the [potential funding block IDs for a given scripthash](https://github.com/romanz/electrs/blob/master/doc/schema.md#transaction-outputs-index-funding). Then, it'll scan the block's transaction outputs to make sure there is no false positives.

I would suggest adding this RPC to the [Electrum Protocol 2.0 PR](https://github.com/spesmilo/electrumx/pull/90).
